### PR TITLE
[RLlib] Minimal ParamNoise PR.

### PR DESCRIPTION
**Pull request by sven1977**
ParameterNoise Exploration added. Nothing else should have changed. Algos/Policies still behave the exact same as before and do NOT make use of the new class. The latter will require a follow-up PR, which requires many more changes in all Policy classes.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/BUILD [hunk 1]
```diff
@@ -966,7 +966,7 @@ py_test(
 py_test(
     name = "tests/test_eager_support",
     tags = ["tests_dir", "tests_dir_E"],
-    size = "large",
+    size = "enormous",
```
Comment (ericl) [inline rllib/BUILD]: Oof.

File: rllib/evaluation/sampler.py [hunk 2]
```diff
@@ -306,6 +306,14 @@ def get_batch_builder():
     def new_episode():
         episode = MultiAgentEpisode(policies, policy_mapping_fn,
                                     get_batch_builder, extra_batch_callback)
+        # Call each policy's Exploration.on_episode_start method.
+        for p in policies.values():
+            p.exploration.on_episode_start(
+                policy=p,
+                environment=base_env,
+                episode=episode,
+                tf_sess=getattr(p, "_sess", None))
```
Comment (ericl) [inline rllib/evaluation/sampler.py]:
```suggestion
                tf_sess=tf_sess)
```

(It's already available as a local variable here passed to the sampler.)

File: rllib/policy/eager_tf_policy.py [hunk 3]
```diff
@@ -261,15 +261,16 @@ def __init__(self, observation_space, action_space, config):
 
         @override(Policy)
         def postprocess_trajectory(self,
-                                   samples,
+                                   sample_batch,
```
Comment (ericl) [inline rllib/policy/eager_tf_policy.py]: Should call these `rollout_fragment` probably from now on, though this would be a large renaming.

File: rllib/policy/policy.py [hunk 4]
```diff
@@ -214,7 +214,8 @@ def postprocess_trajectory(self,
         Returns:
             SampleBatch: Postprocessed sample batch.
         """
-        return sample_batch
+        # Call our Exploration object's postprocess method.
+        return self.exploration.postprocess_trajectory(self, sample_batch)
```
Comment (ericl) [inline rllib/policy/policy.py]: Instead of doing this in the policy, which would require all custom policies to implement this, why not call policy.exploration.postprocess_trajectory() unconditionally in sample_batch_builder.py:152 (after policy.postprocess_traj?)

File: rllib/evaluation/sampler.py [hunk 5]
```diff
@@ -492,6 +500,14 @@ def _process_observations(base_env, policies, batch_builder_pool,
         if all_done:
             # Handle episode termination
             batch_builder_pool.append(episode.batch_builder)
+            # Call each policy's Exploration.on_episode_end method.
+            for p in policies.values():
+                p.exploration.on_episode_end(
+                    policy=p,
+                    environment=base_env,
+                    episode=episode,
+                    tf_sess=getattr(p, "_sess", None))
```
Comment (ericl) [inline rllib/evaluation/sampler.py]:
```suggestion
                    tf_sess=tf_sess)
```

File: rllib/utils/exploration/exploration.py [hunk 6]
```diff
@@ -32,6 +32,30 @@ def __init__(self,
         self.worker_index = worker_index
         self.framework = check_framework(framework)
 
+    def before_forward_pass(self,
```
Comment (ericl) [inline rllib/utils/exploration/exploration.py]: Can we call this `before_compute_actions`?

File: rllib/utils/exploration/exploration.py [hunk 7]
```diff
@@ -32,6 +32,30 @@ def __init__(self,
...
+                            obs_batch,
+                            *,
+                            state_batches=None,
+                            seq_lens=None,
+                            timestep=None,
+                            explore=None,
+                            tf_sess=None,
+                            **kwargs):
+        """May be overridden to perform preparations before a forward pass.
```
Comment (ericl) [inline rllib/utils/exploration/exploration.py]:
```suggestion
        """Hook for preparations before policy.compute_actions() is called.
```

File: rllib/utils/exploration/exploration.py [hunk 8]
```diff
@@ -64,24 +88,49 @@ def get_exploration_action(self,
...
+            environment (BaseEnv): The environment object we are acting in.
+            episode (int): The number of the episode that is starting.
+            tf_sess (Optional[tf.Session]): In case of tf, the session object.
+        """
+        pass
+
+    def postprocess_trajectory(self, policy, sample_batch, tf_sess=None):
+        """Handles post-processing of done episode trajectories.
+
+        Changes the given batch in place.
```
Comment (ericl) [inline rllib/utils/exploration/exploration.py]:
```suggestion
        Changes the given batch in place. This callback is invoked by the sampler after policy.postprocess_trajectory() is called.
```

File: rllib/utils/exploration/parameter_noise.py [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+        self.episode_started = False
+
+    @override(Exploration)
+    def before_forward_pass(self,
+                            obs_batch,
+                            *,
+                            state_batches=None,
+                            seq_lens=None,
+                            prev_action_batch=None,
+                            prev_reward_batch=None,
```
Comment (ericl) [inline rllib/utils/exploration/parameter_noise.py]: Since we don't need these flags (and I can't think of a common use case for them beyond param noise), can you remove these optional args for now? This is an internal-ish developer API so it is ok to make changes to add more fields as needed in the future.

File: rllib/utils/exploration/parameter_noise.py [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+        # Remove stored noise from weights (only if currently noisy).
+        if self.weights_are_currently_noisy:
+            self._remove_noise(tf_sess=tf_sess)
+
+    @override(Exploration)
+    def postprocess_trajectory(self, policy, sample_batch, tf_sess=None):
+        noisy_action_dist = noise_free_action_dist = None
+        # Adjust the stddev depending on the action (pi)-distance.
+        # Also see [1] for details.
+        inputs, dist_class, _ = policy.compute_distribution_inputs(
```
Comment (ericl) [inline rllib/utils/exploration/parameter_noise.py]:
```suggestion
        noisy_action_dist = policy.compute_action_distribution(
```

File: rllib/utils/exploration/parameter_noise.py [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+            action_dist = softmax(inputs)
+        else:  # TODO(sven): Other action-dist cases.
+            raise NotImplementedError
+
+        if self.weights_are_currently_noisy:
+            noisy_action_dist = action_dist
+        else:
+            noise_free_action_dist = action_dist
+
+        inputs, dist_class, _ = policy.compute_distribution_inputs(
```
Comment (ericl) [inline rllib/utils/exploration/parameter_noise.py]:
```suggestion
        no_noise_action_dist = policy.compute_action_distribution(
```

File: rllib/utils/exploration/parameter_noise.py [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+                        shape=noise.shape,
+                        stddev=self.stddev,
+                        dtype=tf.float32)))
+        return tf.group(*added_noises)
+
+    def _sample_new_noise_and_add(self, *, tf_sess=None, override=False):
+        if self.framework == "tf" and not tf.executing_eagerly():
+            if override and self.weights_are_currently_noisy:
+                tf_sess.run(self.tf_remove_noise_op)
+            tf_sess.run(self.tf_sample_new_noise_and_add_op)
```
Comment (ericl) [inline rllib/utils/exploration/parameter_noise.py]: Is it possible to use the `utils/make_tf_callable` decorator to avoid having to call different code for eager vs graph?

File: rllib/utils/exploration/parameter_noise.py [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+
+        if not self.weights_are_currently_noisy:
+            noisy_action_dist = action_dist
+        else:
+            noise_free_action_dist = action_dist
+
+        # Categorical case (e.g. DQN).
+        if dist_class is Categorical:
+            # Calculate KL-divergence (DKL(clean||noisy)) according to [2].
+            kl_divergence = np.nanmean(
```
Comment (ericl) [inline rllib/utils/exploration/parameter_noise.py]: Why not use `action_dist.kl(other_action_dist)`?

I think these computations should be abstracted away in the action dist class itself, so that we don't need to peek at the "type" of the distribution and do different logic depending on that.

File: rllib/utils/exploration/parameter_noise.py [hunk 14]
```diff
@@ -0,0 +1,386 @@
...
+        inputs, dist_class, _ = policy.compute_distribution_inputs(
+            obs_batch=sample_batch[SampleBatch.CUR_OBS],
+            # TODO(sven): What about state-ins and seq-lens?
+            prev_action_batch=sample_batch.get(SampleBatch.PREV_ACTIONS),
+            prev_reward_batch=sample_batch.get(SampleBatch.PREV_REWARDS),
+            explore=not self.weights_are_currently_noisy)
+
+        # Categorical case (e.g. DQN).
+        if dist_class is Categorical:
+            action_dist = softmax(inputs)
```
Comment (ericl) [inline rllib/utils/exploration/parameter_noise.py]: Shouldn't this be `Categorical(inputs)`?

**Review by ericl — CHANGES_REQUESTED** [review]

File: rllib/utils/exploration/exploration.py (hunk 6 above)
Comment (ericl) [reply rllib/utils/exploration/exploration.py]: Also, could you make the changes to the sampler to call this hook? I think you only need to change sampler.py::_do_policy_eval.

File: rllib/policy/policy.py (hunk 4 above)
Comment (sven1977) [reply rllib/policy/policy.py]: +1

Comment (sven1977) [reply rllib/policy/policy.py]: done

File: rllib/utils/exploration/parameter_noise.py (hunk 13 above)
Comment (sven1977) [reply rllib/utils/exploration/parameter_noise.py]: It's happening off-graph. Our distributions are not tested for framework=None (have no numpy implementations).

File: rllib/utils/exploration/parameter_noise.py (hunk 12 above)
Comment (sven1977) [reply rllib/utils/exploration/parameter_noise.py]: I'll check...

File: rllib/utils/exploration/parameter_noise.py (hunk 13 above)
Comment (ericl) [reply rllib/utils/exploration/parameter_noise.py]: I see, is this a long term TODO then? It seems that ideally we would compute the kl using the action dist objects.

File: rllib/utils/exploration/exploration.py (hunk 6 above)
Comment (sven1977) [reply rllib/utils/exploration/exploration.py]: I moved it into _build_compute_actions. I think that's better. Then we don't have to change Policy.compute_actions anymore.

File: rllib/utils/exploration/parameter_noise.py (hunk 13 above)
Comment (ericl) [reply rllib/utils/exploration/parameter_noise.py]: Actually, why not do this on-graph?

File: rllib/utils/exploration/parameter_noise.py (hunk 9 above)
Comment (sven1977) [reply rllib/utils/exploration/parameter_noise.py]: done.

File: rllib/utils/exploration/parameter_noise.py (hunk 10 above)
Comment (sven1977) [reply rllib/utils/exploration/parameter_noise.py]: Hmm, this is not good. The Policy.compute_action_distribution method was not even part of this PR. Let me see, whether using this method can be avoided for now entirely.

File: rllib/utils/exploration/parameter_noise.py (hunk 13 above)
Comment (sven1977) [reply rllib/utils/exploration/parameter_noise.py]: Absolutely! But yeah, thus far, all distributions, torch/tf, were always in graph, so we cannot use those unless we add framework=None support to all of them.

File: rllib/utils/exploration/parameter_noise.py (hunk 14 above)
Comment (sven1977) [reply rllib/utils/exploration/parameter_noise.py]: Same as above (distributions are either tf/torch, not numpy).

File: rllib/utils/exploration/parameter_noise.py (hunk 10 above)
Comment (sven1977) [reply rllib/utils/exploration/parameter_noise.py]: It's only used in ParameterNoise::post process_traj. I'll rename and re-purpose it as you suggested, but it'll not be used (actually called anywhere) till the next PR.

File: rllib/agents/trainer.py [hunk 15]
```diff
@@ -419,7 +419,8 @@ def __init__(self, config=None, env=None, logger_creator=None):
         config = config or {}
 
         if tf and config.get("eager"):
-            tf.enable_eager_execution()
+            if not tf.executing_eagerly():
+                tf.enable_eager_execution()
```
Comment (ericl) [inline rllib/agents/trainer.py]: This is idempotent so you don't need to check it.

File: rllib/evaluation/sample_batch_builder.py [hunk 16]
```diff
@@ -151,6 +152,11 @@ def postprocess_batch_so_far(self, episode):
                     "from a single trajectory.", pre_batch)
             post_batches[agent_id] = policy.postprocess_trajectory(
                 pre_batch, other_batches, episode)
+            # Call the Policy's Exploration's postprocess method.
+            if hasattr(policy, "exploration") and isinstance(
+                    policy.exploration, Exploration):
```
Comment (ericl) [inline rllib/evaluation/sample_batch_builder.py]: Don't we require all policies have exploration defined?

Comment (ericl) [reply rllib/evaluation/sample_batch_builder.py]: (We seem to assume this in sampler.py below)

File: rllib/policy/tf_policy.py [hunk 17]
```diff
@@ -533,6 +533,16 @@ def _build_compute_actions(self,
             raise ValueError(
                 "Must pass in RNN state batches for placeholders {}, got {}".
                 format(self._state_inputs, state_batches))
+
+        # Call the exploration before_forward_pass hook.
+        self.exploration.before_forward_pass(
```
Comment (ericl) [inline rllib/policy/tf_policy.py]: You should do this in sampler.py instead of here. Otherwise, we have to change this for each policy.

File: rllib/utils/exploration/parameter_noise.py (hunk 13 above)
Comment (ericl) [reply rllib/utils/exploration/parameter_noise.py]: Ok, can you leave a TODO here so the reader knows why the code is currently in this form?

**Review by ericl — COMMENTED** [review]
Close...

File: rllib/utils/exploration/parameter_noise.py (hunk 12 above)
Comment (sven1977) [reply rllib/utils/exploration/parameter_noise.py]: I would need the session already in the `make_tf_callable` decorator.

File: rllib/agents/trainer.py (hunk 15 above)
Comment (sven1977) [reply rllib/agents/trainer.py]: I do remember this broke some test recently. Forgot, which one and in which situation.

File: rllib/policy/tf_policy.py (hunk 17 above)
Comment (sven1977) [reply rllib/policy/tf_policy.py]: What about TFPolicy.compute_actions? We would have to do it there then as well.

File: rllib/evaluation/sample_batch_builder.py (hunk 16 above)
Comment (sven1977) [reply rllib/evaluation/sample_batch_builder.py]: Do we? Originally, we said yes, but then we discussed briefly last week that this may not be the case.
In our doc, it says that every policy has one.
Ok, let's assume that this is the case. Every Policy has an Exploration.

File: rllib/policy/tf_policy.py (hunk 17 above)
Comment (sven1977) [reply rllib/policy/tf_policy.py]: Moved into sampler AND TFPolicy.compute_actions.

File: rllib/evaluation/sample_batch_builder.py (hunk 16 above)
Comment (sven1977) [reply rllib/evaluation/sample_batch_builder.py]: Done.

File: rllib/utils/exploration/parameter_noise.py (hunk 13 above)
Comment (sven1977) [reply rllib/utils/exploration/parameter_noise.py]: Done.

File: rllib/policy/tf_policy.py (hunk 17 above)
Comment (ericl) [reply rllib/policy/tf_policy.py]: Ah I see.

File: rllib/evaluation/sample_batch_builder.py (hunk 16 above)
Comment (ericl) [reply rllib/evaluation/sample_batch_builder.py]: Yeah not sure... it seems reasonable if everyone calls super().__init__() on policy.

Comment (sven1977) [reply rllib/evaluation/sample_batch_builder.py]: Totally. The problem is that everyone is doing that at a different position in their (custom) c'tors, sometimes at the very end (not-preferred) sometimes at the beginning (preferred).
That's why it's so important to do the super() call as early as possible. Then the rest of the constructor can assume that the super props are there already. This is not the case e.g. for DynamicTFPolicy and e.g. DDPG.  We should fix this in the "Policy cleanup" PR ahead.

File: rllib/evaluation/sampler.py [hunk 18]
```diff
@@ -558,15 +574,21 @@ def _do_policy_eval(tf_sess, to_eval, policies, active_episodes):
...
-            rnn_in_cols = _to_column_format(rnn_in)
+
+            obs_batch = [t.obs for t in eval_data]
+            state_batches = _to_column_format(rnn_in)
+
+            # Call the exploration before_compute_actions hook.
+            policy.exploration.before_compute_actions(
+                obs_batch=obs_batch,
+                timestep=policy.global_timestep,
+                tf_sess=policy.get_session())
```
Comment (ericl) [inline rllib/evaluation/sampler.py]: Now that you mentioned we need to call before_compute_actions() in the policy anyways, it makes sense to do it in just the policy classes and never the sampler. Sorry about that, can you revert this change?

File: rllib/policy/tf_policy.py [hunk 19]
```diff
@@ -253,6 +253,13 @@ def compute_actions(self,
                         timestep=None,
                         **kwargs):
         explore = explore if explore is not None else self.config["explore"]
+
+        # Call the exploration before_compute_actions hook.
+        self.exploration.before_compute_actions(
+            obs_batch=obs_batch,
+            timestep=self.global_timestep,
+            tf_sess=self.get_session())
```
Comment (ericl) [inline rllib/policy/tf_policy.py]: Should also do in _build_compute_actions()

File: rllib/utils/exploration/parameter_noise.py [hunk 20]
```diff
@@ -0,0 +1,386 @@
...
+
+        # Store the default setting for `explore`.
+        self.default_explore = policy_config["explore"]
+        # Whether we need to call `self._delayed_on_episode_start` before
+        # the forward pass.
+        self.episode_started = False
+
+    @override(Exploration)
+    def before_compute_actions(self,
+                               obs_batch,
```
Comment (ericl) [inline rllib/utils/exploration/parameter_noise.py]: Can you also remove obs_batch for now?

File: rllib/utils/exploration/parameter_noise.py (hunk 12 above)
Comment (ericl) [reply rllib/utils/exploration/parameter_noise.py]: I think you can get that with `get_default_session()` in the exploration constructor, which should always be set correctly when the policy is being constructed.

File: rllib/evaluation/sampler.py (hunk 18 above)
Comment (sven1977) [reply rllib/evaluation/sampler.py]: Done

File: rllib/policy/tf_policy.py (hunk 19 above)
Comment (sven1977) [reply rllib/policy/tf_policy.py]: Done

File: rllib/utils/exploration/parameter_noise.py (hunk 20 above)
Comment (sven1977) [reply rllib/utils/exploration/parameter_noise.py]: Done

File: rllib/utils/exploration/parameter_noise.py (hunk 12 above)
Comment (sven1977) [reply rllib/utils/exploration/parameter_noise.py]: I don't think this will work. What if I need to use with `tf.control_dependencies()` in a  `make_tf_callable` func? Eager would complain, I think.

Comment (sven1977) [reply rllib/utils/exploration/parameter_noise.py]: Let me try ...

**Review by ericl — APPROVED** [review]
LGTM

**Comment by sven1977** [comment]
@ericl  👍 Tests ok, please merge.
